crates/biome_js_analyze/src/analyzers/nursery/no_done_callback.rs

```diff
@@ -0,0 +1,168 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, Ast, Rule, RuleDiagnostic, RuleSource, RuleSourceKind,
+};
+use biome_console::markup;
+use biome_js_syntax::{
+    AnyJsArrowFunctionParameters, AnyJsBinding, AnyJsExpression, JsCallExpression, JsParameters,
+    JsTemplateExpression,
+};
+use biome_rowan::{AstNode, TextRange};
+
+declare_rule! {
+    /// This rule checks the function parameter of hooks & tests for use of the done argument, suggesting you return a promise instead.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// beforeEach(done => {
+    ///     // ...
+    /// });
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// test('myFunction()', done => {
+    ///     // ...
+    /// });
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// beforeEach(async () => {
+    ///     // ...
+    /// });
+    /// ```
+    ///
+    /// ```js
+    /// test('myFunction()', () => {
+    ///     expect(myFunction()).toBeTruthy();
+    /// });
+    /// ```
+    ///
+    pub NoDoneCallback {
+        version: "next",
+        name: "noDoneCallback",
+        recommended: true,
+        source: RuleSource::EslintJest("no-done-callback"),
+        source_kind: RuleSourceKind::Inspired,
+    }
+}
+
+impl Rule for NoDoneCallback {
+    type Query = Ast<JsCallExpression>;
+    type State = TextRange;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+
+        let callee = node.callee().ok()?;
+
+        let mut is_test_each = false;
+        let callee_member = callee.get_callee_member_name();
+        if let Some(member_name) = callee_member {
+            if member_name.text_trimmed() == "each" {
+                is_test_each = true;
+            }
+        }
+
+        if is_test_each && !JsTemplateExpression::can_cast(callee.syntax().kind()) {
+            return None;
+        }
+
+        let arguments = &node.arguments().ok()?;
+        let callee_name = callee.get_callee_object_name()?;
+
+        let argument = match callee_name.text_trimmed() {
+            "after" | "afterAll" | "afterEach" | "before" | "beforeAll" | "beforeEach" => {
+                let argument_by_index = arguments.get_arguments_by_index([0]);
+                let top_arg = argument_by_index.first()?;
+                top_arg.clone()
+            }
+            "it" | "test" => {
+                // for test.each() and test() we want the second argument
+                let argument_by_index = arguments.get_arguments_by_index([1]);
+                let top_arg = argument_by_index.first()?;
+                top_arg.clone()
+            }
+            _ => None,
+        };
```

Comment:
imo; The difference of the cases is only returned index, so it can be like this. 
```suggestion
        let argument_index = match callee_name.text_trimmed() {
            "after" | "afterAll" | "afterEach" | "before" | "beforeAll" | "beforeEach" => 0,
            "it" | "test" => 1, // for test.each() and test() we want the second argument
            _ => return None,
        };
        let argument = arguments
            .get_arguments_by_index([argument_index])
            .first()?
            .clone();
```